src/tests/catwalk_tests/test_storage.py

```diff
@@ -185,29 +219,30 @@ def test_as_of_dates_entity_date_index(project_storage):
         "feature_one": [0.5, 0.6, 0.5, 0.6],
         "feature_two": [0.5, 0.6, 0.5, 0.6],
         "as_of_date": ["2016-01-01", "2016-01-01", "2017-01-01", "2017-01-01"],
+        "label": [1, 0, 1, 0]
     }
     matrix_store = CSVMatrixStore(project_storage, [], "test")
-    matrix_store.matrix = pd.DataFrame.from_dict(data).set_index(
+    matrix_store.matrix_with_labels = pd.DataFrame.from_dict(data).set_index(
         ["entity_id", "as_of_date"]
     )
-    matrix_store.metadata = {"indices": ["entity_id", "as_of_date"]}
+    matrix_store.metadata = {"indices": ["entity_id", "as_of_date"], "label_name": "label"}
 
     assert matrix_store.as_of_dates == ["2016-01-01", "2017-01-01"]
 
 
 def test_s3_save():
     with mock_s3():
-
         client = boto3.client("s3")
         client.create_bucket(Bucket="fake-matrix-bucket", ACL="public-read-write")
-        example = next(matrix_stores())
-        project_storage = ProjectStorage("s3://fake-matrix-bucket")
-
-        tosave = CSVMatrixStore(project_storage, [], "test")
-        tosave.matrix = example.matrix
-        tosave.metadata = example.metadata
-        tosave.save()
-
-        tocheck = CSVMatrixStore(project_storage, [], "test")
-        assert tocheck.metadata == example.metadata
-        assert tocheck.matrix.to_dict() == example.matrix.to_dict()
+        for example in matrix_stores():
+            project_storage = ProjectStorage("s3://fake-matrix-bucket")
+
+            tosave = CSVMatrixStore(project_storage, [], "test")
+            tosave.metadata = example.metadata
+            tosave.matrix_with_labels = example.matrix_with_labels
+            tosave.save()
+
+            tocheck = CSVMatrixStore(project_storage, [], "test")
+            assert tocheck.metadata == example.metadata
+            assert tocheck.design_matrix.to_dict() == example.design_matrix.to_dict()
+            break
```

Comment:
When I do this, the test fails because the temporary directory that was yielded from matrix_stores(). This was surprising to me: I thought that because this was yielded and the temporary directory context manager would still be in effect....but sure enough, it breaks the test.

This wasn't a problem earlier, because the caching was in effect, but now we're testing without caching so that directory needs to be there.

It works if I make the matrix_stores into a context manager. This might be clearer, although doesn't help with indentation.

Reply:
Even better: use the loop, but just skip the HDF ones with continue

Reply:
Yeah, it makes sense that if you don't exhaust the generator, it will never clean up the temporary directory.

Of course this can be resolved in the tests that use this. But that suggests to me that this fixture/helper is doing too much. It can be more opt-in, more flexible.

Or, yeah, if you really want to have a single function that by default does all these things, even though sometimes you don't need all of them, then it can be written as a context manager, or simply as a pytest fixture. Either way you're assured that it will complete, regardless of if the caller consumes all of what it returns/yields.